worker/task.go

```diff
@@ -0,0 +1,201 @@
+	context         *runtime.TaskContext
+	plugin          plugins.TaskPlugin
+	log             *logrus.Entry
+	payload         interface{}
+}
+
+func (t *TaskRun) Prepare(engine engines.Engine) (<-chan engines.SandboxBuilder, <-chan error) {
+	out := make(chan engines.SandboxBuilder, 1)
+	errc := make(chan error, 1)
+	go func() {
+		defer close(errc)
+		t.log.Debug("Preparing task run")
+		err := t.plugin.Prepare(t.context)
+
+		if err != nil {
+			t.log.WithField("error", err.Error()).Warn("Could not prepare task plugins")
+			errc <- err
+			close(out)
+			return
+		}
+
+		sandboxBuilder, err := engine.NewSandboxBuilder(engines.SandboxOptions{
+			TaskContext: t.context,
+			Payload:     t.payload,
+		})
+		if err != nil {
+			t.log.WithField("error", err.Error()).Warn("Could not create sandbox builder")
+			errc <- err
+			close(out)
+			return
+		}
+
+		out <- sandboxBuilder
+
+	}()
+	return out, errc
+}
+
+func (t *TaskRun) Build(in <-chan engines.SandboxBuilder, done <-chan struct{}, inerrc <-chan error) (<-chan engines.SandboxBuilder, <-chan error) {
+	out := make(chan engines.SandboxBuilder)
+	errc := make(chan error, 1)
+
+	go func() {
+		defer close(errc)
+		sb := <-in
+		if sb == nil {
+			errc <- <-inerrc
+			close(out)
+			return
+		}
+		t.log.Debug("Building task run")
+		err := t.plugin.BuildSandbox(sb)
+		if err != nil {
+			t.log.WithField("error", err.Error()).Warn("Could not build build sandbox")
+			errc <- err
+			close(out)
+			return
+		}
+		out <- sb
+	}()
+	return out, errc
+}
+
+func (t *TaskRun) Run(in <-chan engines.SandboxBuilder, done <-chan struct{}, inerrc <-chan error) (<-chan engines.ResultSet, <-chan error) {
+	out := make(chan engines.ResultSet)
+	errc := make(chan error, 1)
+
+	go func() {
+		defer func() {
+			close(errc)
+			close(out)
+		}()
+		sb := <-in
+		if sb == nil {
+			errc <- <-inerrc
+			close(out)
+			return
+		}
+		t.log.Debug("Running task")
+		sandbox, err := sb.StartSandbox()
+		if err != nil {
+			t.log.WithField("error", err.Error()).Warn("Could not start sandbox")
+			errc <- err
+			close(out)
+			return
+		}
+
+		err = t.plugin.Started(sandbox)
+		if err != nil {
+			t.log.WithField("error", err.Error()).Warn("Could not properly start sandbox")
+			errc <- err
+			close(out)
+			return
+		}
+
+		finished := make(chan struct{})
+		go func() {
+			select {
+			case <-done:
+				sandbox.Abort()
+			case <-finished:
+				return
+			}
+		}()
+
+		result, err := sandbox.WaitForResult()
+		close(finished)
+		if err != nil {
+			t.log.WithField("error", err.Error()).Warn("Error when waiting for result set from sandbox")
+			errc <- err
+		}
+		out <- result
+	}()
+	return out, errc
+}
+
+func (t *TaskRun) Stop(in <-chan engines.ResultSet, done <-chan struct{}, inerrc <-chan error) (<-chan bool, <-chan error) {
+	out := make(chan bool, 1)
+	errc := make(chan error, 1)
+
+	go func() {
+		defer func() {
+			close(errc)
+			close(out)
+		}()
+		r := <-in
```

Comment:
wouldn't you need to handled the case where `r` is `nil` ? because there was an error and the input channel was closed..